[SwiftUI] Add a WebView initializer that takes a URL parameter

pxlcoder opened the pull request:
#### 81948b562879648f842aa5be8babc6f501eb4a25
<pre>
[SwiftUI] Add a WebView initializer that takes a URL parameter
<a href="https://bugs.webkit.org/show_bug.cgi?id=291962">https://bugs.webkit.org/show_bug.cgi?id=291962</a>
<a href="https://rdar.apple.com/149246161">rdar://149246161</a>

Reviewed by Richard Robinson.

Make it easier for clients to load a URL, without having to create/configure a
`WebPage`.

In order to achieve this a `WebPage` is created on behalf of the client, and
a `URLRequest` with the specified `URL` is loaded whenever the url changes.
Additionally, note that switching to a nil URL from a non-nil URL should have
no effect.

To avoid recreating the `WebPage` each time SwiftUI recreates the `WebView`,
`Storage` is used to keep the implicitly created `WebPage` as a state variable.
A `URL` may only be extracted when the state variable is in use.

* Source/WebKit/_WebKit_SwiftUI/API/WebView.swift:
(WebView.page):
(WebView.body):
(Storage.webPage):
(Storage.url):

Canonical link: <a href="https://commits.webkit.org/294074@main">https://commits.webkit.org/294074@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d969a4625e085046976d27af07370ef327ef5667

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100663 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20315 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10614 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105800 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51251 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20623 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28789 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76654 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33687 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103670 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15832 "Found 1 new test failure: fast/forms/ios/file-upload-panel-dismiss-when-view-removed-from-window.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90924 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57008 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15641 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50627 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85552 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9006 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108155 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27781 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20401 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85608 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28144 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87125 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85149 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29849 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7577 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21769 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16388 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27716 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32969 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27527 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30845 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29085 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/_WebKit_SwiftUI/API/WebView.swift:
```diff
@@ -24,6 +24,25 @@
 public import SwiftUI
 public import WebKit
 
+private enum StateWithURLOrWebPage: DynamicProperty {
```
```
extension WebView {
     private enum Storage 
}
```

I would do something like this instead, it's more conventional

rr-codes commented on Source/WebKit/_WebKit_SwiftUI/API/WebView.swift:
```diff
@@ -36,13 +55,31 @@ public struct WebView: View {
     }
 
-    let page: WebPage
+    /// Create a new WebView with the specified URL.
```
can you modify this comment to be

```
+    /// Create a new WebView with the specified URL.
+    ///
+    /// For example, you can create a WebView that displays one of two URLs depending on the state of a toggle:
+    ///
+    ///     struct URLView: View {
+    ///         @State private var url: URL? = nil
+    ///         @State private var toggle = false
+    ///
+    ///         var body: some View {
+    ///             VStack {
+    ///                 Button("Toggle") {
+    ///                     toggle.toggle()
+    ///                 }
+    ///                 WebView(initialURL: url)
+    ///             }
+    ///             .onChange(of: toggle, initial: true) {
+    ///                 url = toggle ? URL(string: "https://www.webkit.org") : URL(string: "https://www.apple.com")
+    ///             }
+    ///         }
+    ///     }
+    ///
+    /// - Parameter initialURL: The URL to display in the view. If this value is non-nil or changes to become a non-nil value, the new URL is loaded into the view`
```

pxlcoder replied:
Yes, except `initialURL` -> `url` 🙂

rr-codes commented on Source/WebKit/_WebKit_SwiftUI/API/WebView.swift:
```diff
@@ -36,13 +55,31 @@ public struct WebView: View {
+    private let storage: StateWithURLOrWebPage
+
+    var page: WebPage {
+        return storage.webPage
```
no need for `return`

rr-codes commented on Source/WebKit/_WebKit_SwiftUI/API/WebView.swift:
```diff
@@ -24,6 +24,25 @@
+    var webPage: WebPage {
+        switch self {
+        case .state(let state, _): state.wrappedValue
+        case .webPage(let webPage): webPage
```
```suggestion
        case let .state(state, _): state.wrappedValue
        case let .webPage(webPage): webPage
```

this is a style nit, but it's usually preferred to have the let at the beginning for consistency

rr-codes commented on Source/WebKit/_WebKit_SwiftUI/API/WebView.swift:
```diff
@@ -24,6 +24,25 @@
+
+    var url: URL? {
+        switch self {
+        case .state(_, let url): url
```
ditto about let

rr-codes commented on Source/WebKit/_WebKit_SwiftUI/API/WebView.swift:
```diff
@@ -36,13 +55,31 @@ public struct WebView: View {
     /// - Parameter page: The ``WebPage`` that should be associated with this ``WebView``. It is a programming error to create multiple ``WebView``s with the same ``WebPage``.
     public init(_ page: WebPage) {
-        self.page = page
+        storage = .webPage(page)
```
in initializers, we prefer using `self.` explicitly

rr-codes commented on Source/WebKit/_WebKit_SwiftUI/API/WebView.swift:
```diff
@@ -152,4 +189,23 @@ extension WebView {
         public let linkURL: URL?
     }
+
+    private enum Storage: DynamicProperty {
```
oh this should probably go either in an extension of WebView, or at the top of WebView (nested types usually do one of these)

axiixc replied:
I think a `// MARK: Nested Types` could also work. Although I personally think extensions-as-organization are overrated, especially for protocol conformances, since they don’t surface helpful information in Xcode even.

pxlcoder committed:
[SwiftUI] Add a WebView initializer that takes a URL parameter
https://bugs.webkit.org/show_bug.cgi?id=291962
rdar://149246161

Reviewed by Richard Robinson.

Make it easier for clients to load a URL, without having to create/configure a
`WebPage`.

In order to achieve this a `WebPage` is created on behalf of the client, and
a `URLRequest` with the specified `URL` is loaded whenever the url changes.
Additionally, note that switching to a nil URL from a non-nil URL should have
no effect.

To avoid recreating the `WebPage` each time SwiftUI recreates the `WebView`,
`Storage` is used to keep the implicitly created `WebPage` as a state variable.
A `URL` may only be extracted when the state variable is in use.

* Source/WebKit/_WebKit_SwiftUI/API/WebView.swift:
(WebView.page):
(WebView.body):
(Storage.webPage):
(Storage.url):

Canonical link: https://commits.webkit.org/294074@main

axiixc commented on Source/WebKit/_WebKit_SwiftUI/API/WebView.swift:
```diff
@@ -36,13 +36,46 @@ public struct WebView: View {
+    ///     }
+    ///
+    /// - Parameter url: The URL to display in the view. If this value is non-nil or changes to become a non-nil value, the new URL is loaded into the view.
+    public init(url: URL?) {
```
I’ll admit, I’ve still got some reservations about this as API. Less from a can we do it but a should we do it perspective though.

I think fact that this will navigate any time the URL changes is going to present problems for consistency in more complex cases. Like how do we honor this when there is a `beforeunload` style alert? What happens if the load fails? I worry that this argument breaks the convention of views not really ignoring their inputs (if something `alert(title:)` ever showed a different title, that’d be really weird).

Some of those issues go away if this becomes an `init(initialURL:)`, although that does introduce other oddities, like “how _does_ this ever update?”.

Taking in a `Binding<URL?>` has probably the most consistent semantics, at the cost of requiring the caller to allocate `@State` (or be bad and use `Binding.constant(_:)`, but then that’s on them).

axiixc commented on Source/WebKit/_WebKit_SwiftUI/API/WebView.swift:
```diff
@@ -36,13 +36,46 @@ public struct WebView: View {
+    ///
+    /// - Parameter url: The URL to display in the view. If this value is non-nil or changes to become a non-nil value, the new URL is loaded into the view.
+    public init(url: URL?) {
+        self.storage = .state(State(initialValue: WebPage()), url)
```
Nit, but I’d avoid `init(initialValue:)`. It’s an old spelling that ended up being discarded from the `@propertyWrapper` API, but got left in for source compatibility. It also definitely eagerly invokes the initial value, which you’d want to avoid as much as possible (although I’m not actually sure who calls the `@autoclosure` version from the swiftinterface…).

```swift
  @available(iOS 17.0, macOS 14.0, tvOS 17.0, watchOS 10.0, *)
  @usableFromInline
  internal init(wrappedValue thunk: @autoclosure @escaping () -> Value) where Value : AnyObject, Value : Observation.Observable

  @_alwaysEmitIntoClient public init(initialValue value: Value) {
        _value = value
    }
```

axiixc commented on Source/WebKit/_WebKit_SwiftUI/API/WebView.swift:
```diff
@@ -36,13 +36,46 @@ public struct WebView: View {
+                    return
+                }
+
+                storage.webPage.load(URLRequest(url: url))
```
Like the lack of control over this part of the behavior is what worries me that this API may become overly attractive, but then still require refactoring down the line.

axiixc commented on Source/WebKit/_WebKit_SwiftUI/API/WebView.swift:
```diff
@@ -152,4 +185,25 @@ extension WebView {
+}
+
+extension WebView {
+    private enum Storage: DynamicProperty {
```
And we don’t need to like forward `func update()` to the wrapped `State` here, right?

axiixc commented on Source/WebKit/_WebKit_SwiftUI/API/WebView.swift:
```diff
@@ -152,4 +185,25 @@ extension WebView {
+            }
+        }
+
+        var url: URL? {
```
Not `public`, so this is less impactful, but I do worry this might overlaps a little bit with “the webpage’s current url” just from it’s name (when people read the name that is). It’s also incorrect to call this `initialURL` given the behaviors here though, so I’m not exactly sure how to clarify it? ~~`preferredURL`~~?
